traversal/procedure/ProcedureEdge.java

```diff
@@ -870,50 +877,47 @@ public Forwardable<? extends ThingVertex, Order.Asc> branch(
                             GraphManager graphMgr, Vertex<?, ?> fromVertex, Traversal.Parameters params
                     ) {
                         assert fromVertex.isThing();
-                        Forwardable<? extends AttributeVertex<?>, Order.Asc> iter;
-                        com.vaticle.typedb.core.traversal.predicate.Predicate.Value<?, ?> eq = null;
                         ThingVertex owner = fromVertex.asThing();
-                        if (to.props().hasIID()) iter = branchToIID(graphMgr, params, owner);
-                        else {
-                            eq = iterate(to.props().predicates()).filter(p -> p.operator().equals(EQ)).firstOrNull();
-                            iter = branchToTypes(graphMgr, params, owner, eq);
+                        if (to.props().hasIID()) {
+                            Optional<AttributeVertex<?>> attributeVertex = branchToIID(graphMgr, params, owner);
+                            if (attributeVertex.isPresent()) {
+                                return to.iterateAndFilterPredicates(attributeVertex.get(), params, ASC);
+                            } else return emptySorted(ASC);
+                        } else {
+                            Optional<Value<?, ?>> eq = iterate(to.props().predicates()).filter(p -> p.operator().equals(EQ)).first();
+                            if (eq.isPresent()) {
+                                return to.iterateAndFilterPredicates(branchToEq(graphMgr, params, owner, eq.get()), params, ASC );
+                            } else {
+                                return to.mergeAndFilterPredicatesOnVertices(graphMgr, branchToTypes(graphMgr, owner), params, ASC);
```

Comment:
we can see here that the `to` vertex either filters a single vertex, and list of vertices, or merges and filters a bunch of vertex iterators